packages/gatsby/src/schema/graphql-engine/bundle-webpack.ts

```diff
@@ -38,69 +50,238 @@ function getApisToRemoveForQueryEngine(): Array<GatsbyNodeAPI> {
+      name: `gatsby-internal-packages`,
+      description: `This directory contains internal packages installed by Gatsby used to comply with the current platform requirements`,
+      version: `1.0.0`,
+      private: true,
+      author: `Gatsby`,
+      license: `MIT`,
+      functionsTarget,
+    })
+  }
 }
 
-// lmdb module with prebuilt binaries for our platform
-const lmdbPackage = `@lmdb/lmdb-${process.platform}-${process.arch}`
-
-// Detect if the prebuilt binaries for lmdb have been installed. These are installed under @lmdb and are tied to each platform/arch. We've seen instances where regular installations lack these modules because of a broken lockfile or skipping optional dependencies installs
-function installPrebuiltLmdb(): boolean {
+function getLMDBBinaryFromSiteLocation(
+  lmdbPackageName: string,
+  version: string,
+  functionsTarget: IPlatformAndArch
+): string | undefined {
   // Read lmdb's package.json, go through its optional depedencies and validate if there's a prebuilt lmdb module with a compatible binary to our platform and arch
   let packageJson: PackageJson
   try {
     const modulePath = path
-      .dirname(require.resolve(`lmdb`))
+      .dirname(slash(require.resolve(`lmdb`)))
       .replace(`/dist`, ``)
     const packageJsonPath = path.join(modulePath, `package.json`)
     packageJson = JSON.parse(fs.readFileSync(packageJsonPath, `utf-8`))
   } catch (e) {
     // If we fail to read lmdb's package.json there's bigger problems here so just skip installation
-    return false
+    return undefined
   }
   // If there's no lmdb prebuilt package for our arch/platform listed as optional dep no point in trying to install it
   const { optionalDependencies } = packageJson
-  if (!optionalDependencies) return false
-  if (!Object.keys(optionalDependencies).find(p => p === lmdbPackage))
-    return false
+  if (
+    !Object.keys(optionalDependencies ?? {}).find(p => p === lmdbPackageName)
+  ) {
+    throw new Error(
+      `Target platform/arch for functions execution (${functionsTarget.platform}/${functionsTarget.arch}) is not supported.`
+    )
+  }
+  return getPackageLocationFromRequireContext(
+    slash(require.resolve(`lmdb`)),
+    lmdbPackageName,
+    version
+  )
+}
+
+function getPackageLocationFromRequireContext(
+  location: string,
+  packageName: string,
+  packageVersion?: string
+): string | undefined {
   try {
-    const lmdbRequire = mod.createRequire(require.resolve(`lmdb`))
-    lmdbRequire.resolve(lmdbPackage)
-    return false
+    const requireId = `${packageName}/package.json`
+    const locationRequire = mod.createRequire(location)
+    const packageJsonLocation = slash(locationRequire.resolve(requireId))
+
+    if (packageVersion) {
+      // delete locationRequire.cache[requireId]
+      const { version } = JSON.parse(
+        fs.readFileSync(packageJsonLocation, `utf-8`)
+      )
+      if (packageVersion !== version) {
+        return undefined
```

Comment:
Same comment as before, if we can `''` would allow us to keep the return type strictly `string`